crates/turborepo-lib/src/daemon/server.rs

```diff
@@ -293,31 +279,77 @@ where
         trace!("root watching exited");
         // Clean up the filewatching handle in the event that we never even got
         // started with filewatching. Again, we don't care about the error here.
-        let _ = fw_handle.await;
+        // let _ = fw_handle.await;
         trace!("filewatching handle joined");
         Ok(close_reason)
     }
 }
 
-struct TurboGrpcServiceInner {
+struct TurboGrpcServiceInner<PD> {
     shutdown: mpsc::Sender<()>,
-    watcher_rx: watch::Receiver<Option<Arc<FileWatching>>>,
+    file_watching: FileWatching,
     times_saved: Arc<Mutex<HashMap<String, u64>>>,
     start_time: Instant,
     log_file: AbsoluteSystemPathBuf,
+    package_discovery: PD,
 }
 
-impl TurboGrpcServiceInner {
+// we have a grpc service that uses watching package discovery, and where the
+// watching package hasher also uses watching package discovery as well as
+// falling back to a local package hasher
+impl TurboGrpcServiceInner<Arc<WatchingPackageDiscovery>> {
+    pub fn new<PD: Sync + PackageDiscovery + Send + 'static>(
+        package_discovery_backup: PD,
+        repo_root: AbsoluteSystemPathBuf,
+        trigger_shutdown: mpsc::Sender<()>,
+        log_file: AbsoluteSystemPathBuf,
+    ) -> (
+        Self,
+        oneshot::Sender<()>,
+        JoinHandle<Result<(), WatchError>>,
+    ) {
+        let file_watching = FileWatching::new(repo_root.clone(), package_discovery_backup).unwrap();
```

Comment:
Good point. Really, everything can move out (package discovery watcher, and file watcher) when the time comes. Since we are creating the package discovery in here also I think it makes sense but it makes sense to move it out if/when we decide to provide alternative APIs